File: typescript/src/backoff.ts
export interface ReconnectBackoffPolicy {
  readonly baseMs: number;
  readonly maxMs: number;
}

export function reconnectDelayMs(
  attempt: number,
  policy: ReconnectBackoffPolicy,
): number {
  if (!Number.isSafeInteger(attempt) || attempt < 0) {
    throw new RangeError("attempt must be a non-negative safe integer");
  }
  if (!Number.isSafeInteger(policy.baseMs) || policy.baseMs < 0) {
    throw new RangeError("baseMs must be a non-negative safe integer");
  }
  if (!Number.isSafeInteger(policy.maxMs) || policy.maxMs < policy.baseMs) {
    throw new RangeError("maxMs must be a safe integer at least baseMs");
  }
  return Math.min(policy.maxMs, policy.baseMs * 2 ** attempt);
}

File: rust/src/backoff.rs
use crate::{MAX_INTEROPERABLE_INTEGER, RealtimeError};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct ReconnectBackoffPolicy {
    pub base_ms: u64,
    pub max_ms: u64,
}

/// Returns the deterministic exponential delay for a zero-based reconnect attempt.
///
/// # Errors
///
/// Returns [`RealtimeError::InvalidBackoffPolicy`] when `max_ms` is below `base_ms` or either
/// delay exceeds the exact cross-language integer range.
pub fn reconnect_delay_ms(
    attempt: u32,
    policy: ReconnectBackoffPolicy,
) -> Result<u64, RealtimeError> {
    if policy.max_ms < policy.base_ms
        || policy.base_ms > MAX_INTEROPERABLE_INTEGER
        || policy.max_ms > MAX_INTEROPERABLE_INTEGER
    {
        return Err(RealtimeError::InvalidBackoffPolicy);
    }
    if policy.base_ms == 0 || attempt >= u64::BITS {
        return Ok(if policy.base_ms == 0 {
            0
        } else {
            policy.max_ms
        });
    }
    let multiplier = 1_u64 << attempt;
    Ok(policy
        .base_ms
        .checked_mul(multiplier)
        .unwrap_or(policy.max_ms)
        .min(policy.max_ms))
}
